### PR TITLE
WIP - cherry-picking TASK_LOST changes from 0.15

### PR DESCRIPTION
**Pull request by meichstedt**
…timeout.

The lost task is killed after a configurable timeout.

Conflicts:
	docs/docs/command-line-flags.md
	src/main/scala/mesosphere/marathon/MarathonConf.scala
	src/main/scala/mesosphere/marathon/MarathonScheduler.scala
	src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala
	src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala
	src/main/scala/mesosphere/marathon/core/CoreModuleImpl.scala
	src/main/scala/mesosphere/marathon/core/launchqueue/LaunchQueue.scala
	src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala
	src/main/scala/mesosphere/marathon/core/task/jobs/TaskJobsModule.scala
	src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala
	src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerDelegate.scala
	src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala
	src/main/scala/mesosphere/marathon/core/task/update/impl/steps/PostToEventStreamStepImpl.scala
	src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala
	src/test/scala/mesosphere/marathon/MarathonTestHelper.scala
	src/test/scala/mesosphere/marathon/SchedulerActionsTest.scala
	src/test/scala/mesosphere/marathon/api/v2/QueueResourceTest.scala
	src/test/scala/mesosphere/marathon/api/v2/json/MarathonTaskFormatTest.scala
	src/test/scala/mesosphere/marathon/core/launcher/impl/LaunchQueueTestHelper.scala
	src/test/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActorTest.scala
	src/test/scala/mesosphere/marathon/core/task/bus/TaskStatusUpdateTestHelper.scala
	src/test/scala/mesosphere/marathon/core/task/tracker/impl/StatusUpdateActionResolverTest.scala
	src/test/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImplTest.scala
	src/test/scala/mesosphere/marathon/core/task/update/impl/steps/PostToEventStreamStepImplTest.scala
	src/test/scala/mesosphere/marathon/integration/facades/MarathonFacade.scala

### CODE REVIEW

**Comment by meichstedt** [comment]
@aquamatthias PTAL – still missing the simplification of TaskStatus/State handling.

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala [hunk 1]
```diff
@@ -516,8 +516,13 @@ class SchedulerActions(
   /**
     * Make sure the app is running the correct number of instances
     */
+  // FIXME: extract computation into a function that can be easily tested
   def scale(driver: SchedulerDriver, app: AppDefinition): Unit = {
-    val launchedCount = taskTracker.countLaunchedAppTasksSync(app.id)
+    import SchedulerActions._
+
+    // FIXME (merge): there are now 2 filters ...
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: When do we want to address this?

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/PostToEventStreamStepImpl.scala [hunk 2]
```diff
@@ -20,37 +20,41 @@ import scala.concurrent.Future
...
-        mesosStatus.getState match {
-          case Terminated(_) =>
-            postEvent(timestamp, mesosStatus, task)
-          case TaskState.TASK_RUNNING if task.launched.exists(!_.hasStartedRunning) => // staged, not running
-            postEvent(timestamp, mesosStatus, task)
-          case TaskState.TASK_KILLING if task.launched.isDefined =>
-            postEvent(timestamp, mesosStatus, task)
+    import TaskStateOp.MesosUpdate
+
+    // FIXME (merge): it should be clear which task state (old or new) to post
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/steps/PostToEventStreamStepImpl.scala]: Do we want to address this for this PR?

File: src/main/scala/mesosphere/marathon/core/task/Task.scala [hunk 3]
```diff
@@ -38,6 +38,7 @@ sealed trait Task {
     taskId.mesosTaskId
   }
 
+  // FIXME (merge): remove this accessor and add a universal Status to all tasks that includes unconfirmed states
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: When do we want to address this?

File: src/main/scala/mesosphere/marathon/core/task/bus/MarathonTaskStatus.scala [hunk 4]
```diff
@@ -7,9 +7,11 @@ sealed trait MarathonTaskStatus {
   def mesosStatus: Option[TaskStatus]
 }
 
+// FIXME (merge): Merge MarathonTaskStatus and MesosTaskStatus
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/bus/MarathonTaskStatus.scala]: Perhaps it makes sense to have MesosTaskStatus? I see a couple of references. What do you think?

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala [hunk 5]
```diff
@@ -46,25 +46,25 @@ class TaskStatusUpdateProcessorImpl @Inject() (
...
         val taskStr = taskKnownOrNotStr(maybeTask)
         log.info(s"Ignoring ${status.getState} update for $taskStr $taskId")
-        acknowledge(status)
+        acknowledge(status, ack)
     }
   }
 
-  private[this] def acknowledge(taskStatus: MesosProtos.TaskStatus): Future[Unit] = {
-    driverHolder.driver.foreach(_.acknowledgeStatusUpdate(taskStatus))
+  private[this] def acknowledge(taskStatus: MesosProtos.TaskStatus, ack: Boolean = true): Future[Unit] = {
```
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: I think we introduced the ack only for the "invented" TaskStatusUpdate. Do we still need this?

**Comment by aquamatthias** [comment]
Looks good. Nice tests! 
I added a couple of questions - mainly regarding fixmes.

File: src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala (hunk 1 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala]: I've changed this to use `countAppTasksSync` with a combined filter for now.

File: src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala (hunk 5 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/TaskStatusUpdateProcessorImpl.scala]: Good catch; removed.

File: src/main/scala/mesosphere/marathon/core/task/Task.scala (hunk 3 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/Task.scala]: I've removed the fixme for now.
(I'll see if I can safely add the mentioned functionality on top.)

File: src/main/scala/mesosphere/marathon/core/task/update/impl/steps/PostToEventStreamStepImpl.scala (hunk 2 above)
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/update/impl/steps/PostToEventStreamStepImpl.scala]: I don't think so.

**Comment by aquamatthias** [comment]
Thanks!

**Pull request merged**